src/gui/title_bar.cpp

```diff
@@ -0,0 +1,600 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2023-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "title_bar.h"
+
+#include "checks.h"
+#include "control.h"
+#include "cpu.h"
+#include "dosbox.h"
+#include "mapper.h"
+#include "sdlmain.h"
+#include "setup.h"
+#include "string_utils.h"
+#include "support.h"
+#include "video.h"
+
+#include <SDL.h>
+#include <atomic>
+
+CHECK_NARROWING();
+
+// ***************************************************************************
+// Datat types and storage
+// ***************************************************************************
+
+enum class ProgramDisplay : uint8_t { None, Name, Path, Segment };
+enum class VersionDisplay : uint8_t { None, Simple, Detailed };
+
+static struct TitlebarConfig {
+	std::string titlebar_text = {};
+	bool animated_rec         = false;
+	bool show_cycles          = false;
+	bool show_dosbox          = false;
+	ProgramDisplay program    = ProgramDisplay::None;
+	VersionDisplay version    = VersionDisplay::None;
+} config = {};
+
+static struct {
+	bool audio_capture   = false;
+	bool video_capture   = false;
+	bool guest_os_booted = false;
+
+	std::string segment_name   = {};
+	std::string canonical_name = {}; // path + name + extension
+	std::string hint_mouse     = {};
+
+	int num_cycles = 0;
+
+	// Title string variants in case of animated recording mark
+	std::string title_frame_1 = {};
+	std::string title_frame_2 = {};
+
+	SDL_TimerID timer_id = {};
+
+	std::atomic<bool> is_animation_running      = false;
+	std::atomic<bool> animation_phase_alternate = false;
+} state = {};
+
+// Pre-calculated strings, to save some time when generating new title bar text
```

Comment:
Haha, "save some time"... Watch me when the OSD will regenerate the UI at 60-70 FPS using Bezier curves etc 😎